### DISCUSSION TITLE
General Performance Enhancement Thread

### DISCUSSION THREAD

**Discussion by ilikenwf** [vladmandic/sdnext#1102] [General]
We're all still more or less riding the pytorch/diffusers train, but theoretically Tensor cores should be faster...and then there are a litany of other techniques...

Model distillation (which isn't always as nice with output but crazy fast) is an interesting topic, but then I found this repo that seems to have a version of serveral different diffusion techniques combined...

https://github.com/stochasticai/x-stable-diffusion

Because stable-diffusion works and people with funds can just throw big hardware at it, that obscures the what I see to be ineffecient nature of the current technology...

I'm sure there's other repos and papers around we should be looking into. Some of them got proposed and posted on arxiv or reddit, or github, and then got ignored...

**Comment by vladmandic** [comment]
actually, we're riding the traing of pytorch/transformers and amazing a1111 work loosely based on early versions of diffusers.
what we're not using is modern diffusers - and there is a major project ongoing to incorporate them once again as huggingface put a lot of work into diffusers since a1111 project did its implementation.

on the rest, there are ports running sd on tensorrt (and other backends) - the problem with all of those is that it requires a compiled/converted model. ok, models can be compiled - its painful, but its a one time thing. but it also means that the rest of ecosystem is not longer viable. can you you textual inversions? lora? lyco? inpainiting? any type of tiling? any kind of training? pretty much only thing you can do is a) simple generate, b) use upscalers as most of them are completely separate anyhow

so yes, i'm looking how to get there, but i don't want to "enable" tenrsorrt or anything like that if only thing it brings is nice benchmark results, but not usable in real world.

but yeah, if you see a nice optimization in some repo - please report, i'd love to see something we can use!

**Reply by ilikenwf** [reply]
Yeah, makes sense...I thought we did pull in the diffusers in requierments, the master branch, 2.0 not 1.5?

I haven't test with tensorrt yet, part because of compilation, part laziness, many of those are out of date on deps. In the future it would be nice to have this as an option for batch jobs, perhaps...if you're generating hundreds or thousands of images, compile the model to tensor to generate, and then to process later you could use the pytorch stuff with controlnet or whatever.

Personally I think tensors and cuda cores are a stepping stone and something more generic and fast will come around and make all of this stuff outdated eventually anyway, and use a ton less memory to do it...

I'm not sure how to get there either unless we start having Copilot X and gpt assist us...but they're not creative enough yet to do more than point out memory and performance issues that a human would notice most of the time.

**Reply by vladmandic** [reply]
i pulled in diffusers as a requirement as prep work for actually doing some more stuff with them :)

> Personally I think tensors and cuda cores are a stepping stone and something more generic and fast will come around and make all of this stuff outdated eventually anyway, and use a ton less memory to do it...

Yes and no. The problem is that money is in large stuff and that stuff is becoming more and proprietry.
For example, take a look at <https://nvidianews.nvidia.com/news/nvidia-launches-dgx-cloud-giving-every-enterprise-instant-access-to-ai-supercomputer-from-a-browser>

And Google and AWS are doing similar - yes, they offer general GPU/TPU IaaS instances, But big money is in proprietary HW and SW and offering ML-as-a-Service.

And public is being left behind - we are forced to use "secondary" solutions. For example, nVidia tensor cores are not even designed with ML in mind - they are designed performing secondary tasks in gaming (DLSS 2.0 realtime upscaling and DLSS 3.0 frame interpolation sells far more GPUs that enthusiasts playing with StableDiffusion).
Not to mention how little focus is AMD placing on development of ROCm.

Ok, rant off, time to go to sleep :)

**Reply by ilikenwf** [reply]
On the corporate side, though, they are more about forward movement with no concern over cost or efficiency...as open source catches up and we buy cheap ML equipment from yesteryear on ebay, the tides still should turn....and even if not we'll be able to use their own AI's to build stuff that does surpass themselves, and run on what we have available to us...eventually...I hope.

**Comment by ilikenwf** [comment]
Hopefully the person who shared this and then deleted the message can share it again?

https://old.reddit.com/r/StableDiffusion/comments/13o867l/i_made_some_changes_in_automatic1111_sd_webui/

**Reply by ghost** [reply]
I posted it and then I saw that vlad already responded on that reddit thread, so I thought its redundant
